wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginMetaRetriever.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.wsplugins;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.google.common.annotations.Beta;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.CharStreams;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.ws.rs.core.UriBuilder;
+import javax.ws.rs.core.UriBuilderException;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException;
+import org.eclipse.che.api.workspace.server.wsplugins.model.PluginMeta;
+import org.eclipse.che.api.workspace.shared.Constants;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.lang.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Fetches Che tooling plugin objects corresponding to attributes of a workspace config.
+ *
+ * <p>This API is in <b>Beta</b> and is subject to changes or removal.
+ *
+ * @author Oleksander Garagatyi
+ */
+@Beta
+public class PluginMetaRetriever {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PluginMetaRetriever.class);
+  private static final String CHE_PLUGIN_OBJECT_ERROR =
+      "Che plugin '%s:%s' configuration is invalid. %s";
+  private static final String PLUGIN_REGISTRY_PROPERTY = "che.workspace.plugin_registry_url";
+
+  private static final ObjectMapper YAML_PARSER = new ObjectMapper(new YAMLFactory());
+
+  private final UriBuilder pluginRegistry;
+
+  @Inject
+  public PluginMetaRetriever(@Nullable @Named(PLUGIN_REGISTRY_PROPERTY) String pluginRegistry) {
+    if (pluginRegistry == null) {
+      LOG.info(
+          format(
+              "Che tooling plugins feature is disabled - Che plugin registry API endpoint property '%s' is not configured",
+              PLUGIN_REGISTRY_PROPERTY));
+      this.pluginRegistry = null;
+    } else {
+      this.pluginRegistry = UriBuilder.fromUri(pluginRegistry).path("plugins");
```

Comment:
the deploy script is providing URL like
```
 CHE_WORKSPACE_PLUGIN__REGISTRY__URL="${HTTP_PROTOCOL}://${PLUGIN_REGISTRY_ROUTE}/plugins/"
```

and here we append a `/plugins` again

so AFAIK it failed